### ISSUE TITLE
Producer-Stages may introduce Message-Reordering after a failed Send-Request

### ISSUE THREAD

**Issue by cstub** [akka/reactive-kafka#295]
We are using Kafka with reliable messaging semantics, hence we want to guarantee that no message gets lost and that the order of messages per partition is kept as they occur at the source.
With Reactive-Kafka we came across a situation where message-loss or message-reordering may occur due to the mechanics of the `ProducerStage`.

Upon encountering a failed send-request (signaled by an error in the send-callback) the producer in the `ProducerStage` is closed in the `postStop` method but not before all remaining records are flushed. These records may have been submitted to `producer.send` after the failing record. As a result the failing record is dropped and reordering occurs.

According to a [LinkedIn slide deck](https://de.slideshare.net/JiangjieQin/no-data-loss-pipeline-with-apache-kafka-49753844) on preventing data loss with Kafka, the producer should be closed with `producer.close(0)` in the send-callback to ensure that no further records are sent after encountering a failure (see [slide 10]).

With the current implementation of Reactive-Kafka this is not possible, as closing the producer is always preceded by a call to `producer.flush()`. The only possible solution would be to only process one record at a time with `parallelism = 1` and a stream `bufferSize = 1` but this would kill throughput.

Our proposal is to add an option to force-close the producer in the send-callback with `producer.close(0)` for all Producer-APIs that use a separate `KafkaProducer`. This way the user could decide if remaining records should be flushed or dropped in case of an error.

The following ScalaTest reproducer should demonstrate the problem, as failures are provoked for the first few records by setting `max.block.ms = 1`. 
With the requirement of stable message-order, no records should be produced after the first failed record, but the producer does indeed produce a non-deterministic amount of records.

```
import java.util.UUID
import java.util.concurrent.TimeUnit

import akka.NotUsed
import akka.actor.ActorSystem
import akka.kafka.ProducerMessage.Message
import akka.kafka.scaladsl.{Consumer, Producer}
import akka.kafka.{ConsumerSettings, ProducerSettings, Subscriptions}
import akka.stream.ActorMaterializer
import akka.stream.scaladsl.{Keep, Sink, Source}
import akka.stream.testkit.scaladsl.TestSink
import akka.testkit.TestKit
import net.manub.embeddedkafka.{EmbeddedKafka, EmbeddedKafkaConfig}
import org.apache.kafka.clients.consumer.ConsumerConfig
import org.apache.kafka.clients.producer.ProducerRecord
import org.apache.kafka.common.serialization.{ByteArrayDeserializer, ByteArraySerializer, StringDeserializer, StringSerializer}
import org.scalatest.{BeforeAndAfterAll, Matchers, WordSpecLike}

import scala.concurrent.Await
import scala.concurrent.duration._
import scala.language.postfixOps

class Spec extends TestKit(ActorSystem("Spec")) with WordSpecLike with Matchers with BeforeAndAfterAll {
  implicit val mat = ActorMaterializer()(system)
  val bootstrapServers = s"localhost:${EmbeddedKafkaConfig.defaultConfig.kafkaPort}"
  val InitialMsg = "initial message"

  val producerSettings = ProducerSettings(system, new ByteArraySerializer, new StringSerializer)
    .withBootstrapServers(bootstrapServers)

  val consumerSettings = ConsumerSettings(system, new ByteArrayDeserializer, new StringDeserializer)
    .withBootstrapServers(bootstrapServers)
    .withGroupId(s"group-${UUID.randomUUID()}")
    .withProperty(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest")

  override protected def beforeAll(): Unit = {
    EmbeddedKafka.start()
  }

  override def afterAll(): Unit = {
    shutdown(system, 30.seconds)
    EmbeddedKafka.stop()
  }

  "Reactive kafka streams" must {
    "not produce further records after a send-failure" in {
      val topic = s"topic-${UUID.randomUUID()}"

      // create test topic
      val producer = producerSettings.createKafkaProducer()
      producer.send(new ProducerRecord(topic, 0, null: Array[Byte], InitialMsg))
      producer.close(60, TimeUnit.SECONDS)

      // IMPORTANT: ensure that the first few records will fail because the topic metadata request for those records will timeout
      // further records may succeed as the metadata will be available by then
      val failFirstFewRecordsProducerSettings = producerSettings.withProperty("max.block.ms", "1")

      Await.ready(
        Source(1 to 100)
          .map(n => Message(new ProducerRecord(topic, 0, null: Array[Byte], n.toString), NotUsed))
          .viaMat(Producer.flow(failFirstFewRecordsProducerSettings))(Keep.right)
          .runWith(Sink.ignore),
        10.seconds)

      // create consumer probe
      val probe = Consumer.plainSource(consumerSettings, Subscriptions.topics(topic))
        .filterNot(_.value == InitialMsg).map(_.value).runWith(TestSink.probe)

      probe.request(100)

      // since the first record fails no further records should be produced and hence received
      probe.expectNoMsg()
    }
  }
}
```

**Comment by patriknw** [comment]
@cstub Sounds like a good finding. Would you be able to create a pull request?

**Comment by cstub** [comment]
@patriknw Of course. I will summarize my ideas for the changes and keep you posted!

**Comment by cstub** [comment]
I'd propose to extend the `Producer`-API variants which facilitate the non-shared `KafkaProducer` by an optional `FailureStrategy`. This `FailureStrategy` defines how the close-process of the underlying `KafkaProducer` should be handled in case of send-failures. 

The `FailureStrategy` is a `PartialFunction[Throwable, Decision]` where a `Decision` might be to `dropRemainingMessages` (which forces the producer to be closed in the callback) or `flushRemainingMessages` (which gracefully closes the producer in `postStop`).

This `FailureStrategy` is only applied in case of a `SupervisionStrategy` of `Stop,` in order to honor the stream mechanics, otherwhise a force-close of the `KafkaProducer` in the callback would effectively stop message production of the stream even if not desired by the `SupervisionStrategy`.

The default value of the `FailureStrategy` would be to `flushRemainingMessages` to retain the current behaviour.

@patriknw WDYT?

**Comment by cstub** [comment]
In the current PR for this issue the API changes of the initial proposal were discarded in favour 
of a new default failure-behaviour which is only applied if specified by the `Supervision` strategy.

If an error occurs within the send-callback of the `KafkaProducer` and the `Supervision` strategy specifies a behaviour of `Stop` for this error, all unsent messages are discarded as this provides a deterministic failure behaviour. Instead of being uncertain which messages were sent after a failure occurred, all messages after a failed send are discarded.

If errors should be ignored a `Supervision` strategy of `Resume` or `Restart` may be used.

**Comment by ennru** [comment]
My understanding is that this issue was fixed with #318.
Please re-open if I got it wrong.
